test/gtest/common.cpp

```diff
@@ -72,4 +78,65 @@ ScopedEnv::Variable::~Variable()
     }
 }
 
+PortAllocator::PortAllocator()
+    : _concurrent_id(_get_concurrent_id()),
+      _port(_get_first_port(_concurrent_id)) {}
+
+PortAllocator &
+PortAllocator::instance() {
+    static PortAllocator _instance;
+    return _instance;
+}
+
+bool
+PortAllocator::_is_port_available(uint16_t port) {
+    struct sockaddr_in addr = {
+        .sin_family = AF_INET, .sin_port = htons(port), .sin_addr = {.s_addr = INADDR_ANY}};
+
+    const auto sock_fd = socket(AF_INET, SOCK_STREAM, 0);
+    const auto ret = bind(sock_fd, (struct sockaddr *)&addr, sizeof(addr));
+    close(sock_fd);
+    return ret == 0;
+}
+
+uint16_t
+PortAllocator::next_tcp_port() {
+    PortAllocator &instance = PortAllocator::instance();
+    std::lock_guard<std::mutex> lock(instance._mutex);
+    int max_port = MIN_PORT + (instance._concurrent_id + 1) * PORT_RANGE - 1;
+
+    while (!_is_port_available(++instance._port) && (instance._port <= max_port))
+        ;
+
+    if (instance._port >= max_port) {
```

Comment:
maybe wrap around to first port? and fail only if we really scanned all range and not just started from MAX_PORT-1 for example